prefix_metadata.go

```diff
@@ -0,0 +1,221 @@
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package dosa
+
+import (
+	"fmt"
+	"time"
+)
+
+// PrefixMetadata is metadata about a prefix in production (note: we do not track prefixes for dev scopes).
+type PrefixMetadata struct {
+	Entity      `dosa:"primaryKey=(Name)" json:"-"`
+	Name        string     `json:"name"`
+	Owner       string     `json:"owner"` // Owning group name (or the same as Creator)
+	Flags       int64      `json:"flags"`
+	Version     int32      `json:"version"`
+	EntityStr   string     `json:"entity_str,omitempty"` // With ":" separators
+	Cluster     string     `json:"cluster,omitempty"`    // Host DB cluster
+	Creator     string     `json:"creator"`
+	CreatedOn   time.Time  `json:"created_on"`
+	ExpiresOn   *time.Time `json:"expires_on,omitempty"`
+	ExtendCount int32      `json:"extend_count,omitempty"`
+	NotifyCount int32      `json:"notify_count,omitempty"`
+	ReadMaxRPS  int32      `json:"read_max_rps"`
+	WriteMaxRPS int32      `json:"write_max_rps"`
+
+	// This is for convenience only, not stored in the DB:
+	Entities StringSet `dosa:"-" json:"-"`
+}
+
+func (md *PrefixMetadata) String() string {
+	s := fmt.Sprintf("<Scope %s (%s): owner=%s, creator=%s, created=%v", md.Name, Production,
+		md.Owner, md.Creator, md.CreatedOn)
+	if md.ExpiresOn != nil {
+		s += fmt.Sprintf(", expires=%v", *md.ExpiresOn)
+	}
+	if len(md.Entities) > 0 {
+		s += fmt.Sprintf(", entities=%s", md.Entities)
+	}
+	if len(md.Cluster) > 0 {
+		s += fmt.Sprintf(", cluster=%v", md.Cluster)
+	}
+	if md.ExtendCount > 0 {
+		s += fmt.Sprintf(", extended=%d", md.ExtendCount)
+	}
+	if md.NotifyCount > 0 {
+		s += fmt.Sprintf(", notified=%d", md.ExtendCount)
+	}
+	return s + ">"
+}
+
+// PrefixFlagType is a set of prefix flags
+type PrefixFlagType int64
+
+// G E T T E R S
+
+// GetName is the name of the prefix.
+func (md *PrefixMetadata) GetName() string {
+	return md.Name
+}
+
+// GetOwner is the owner of the prefix.
+func (md *PrefixMetadata) GetOwner() string {
+	return md.Owner
+}
+
+// GetType is the type of the prefix.
+func (md *PrefixMetadata) GetType() int32 {
+	return int32(Production)
+}
+
+// GetFlags is the flags of the prefix.
+func (md *PrefixMetadata) GetFlags() int64 {
+	return md.Flags
+}
+
+// GetVersion is the version of the prefix.
+func (md *PrefixMetadata) GetVersion() int32 {
+	return md.Version
+}
+
+// GetEntityStr is the encoded list of all prefixes.
```

Comment:
More useful methods would be `AddEntity()`, `HasEntity() bool`, and `GetEntities() []string`....

Reply:
(and same for prefixes)